bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieImpl.java

```diff
@@ -112,8 +113,8 @@ public class BookieImpl extends BookieCriticalThread implements Bookie {
 
     private int exitCode = ExitCode.OK;
 
-    private final ConcurrentLongHashMap<byte[]> masterKeyCache =
-            ConcurrentLongHashMap.<byte[]>newBuilder().build();
+    private final Cache<Long, byte[]> masterKeyCache = CacheBuilder.newBuilder().
+            expireAfterAccess(60, TimeUnit.MINUTES).build();
```

Comment:
This will also mean that we have to acquire a write lock each time to update the last access time. And The cache implementation has to do the extra work for the eviction.

Reply:
> 

Is there any better implementation to replace guava cache? @merlimat

Reply:
there is Caffeine Cache which claims better perf but I don't know if it behaves any better in this case. 
You can as well switch to expireAfterWrite, make it configurable and let users set it. 
Pulsar has configuration for time based ledger rollover, just have to set it is lower than the expireAfterWrite value in bk.

Reply:
fyi, in both caches a read does not require any locking on the read path